// @logto/next/lib/index.js
var $gpzKJ$logtonode = require("@logto/node");
var $gpzKJ$ironsessionnext = require("iron-session/next");

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}
function $parcel$defineInteropFlag(a) {
  Object.defineProperty(a, '__esModule', {value: true, configurable: true});
}
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}

$parcel$defineInteropFlag(module.exports);

$parcel$export(module.exports, "default", () => $dee2a34a598960d9$export$2e2bcd8739ae039);
$parcel$export(module.exports, "ReservedScope", () => $dee2a34a598960d9$re_export$ReservedScope);
$parcel$export(module.exports, "UserScope", () => $dee2a34a598960d9$re_export$UserScope);


class $f80b46e338398246$export$2e2bcd8739ae039 {
    constructor(request){
        this.request = request;
        this.sessionChanged = false;
    }
    async setItem(key, value) {
        this.request.session[key] = value;
        this.sessionChanged = true;
    }
    async getItem(key) {
        const value = this.request.session[key];
        if (value === undefined) return null;
        return String(value);
    }
    async removeItem(key) {
        this.request.session[key] = undefined;
        this.sessionChanged = true;
    }
    async save() {
        if (!this.sessionChanged) return;
        await this.request.session.save();
        this.sessionChanged = false;
    }
}



class $dee2a34a598960d9$export$2e2bcd8739ae039 {
    constructor(config){
        this.config = config;
        this.handleSignIn = (redirectUri = `${this.config.baseUrl}/api/logto/sign-in-callback`)=>(0, $gpzKJ$ironsessionnext.withIronSessionApiRoute)(async (request, response)=>{
                const nodeClient = this.createNodeClient(request);
                await nodeClient.signIn(redirectUri);
                await this.storage?.save();
                if (this.navigateUrl) response.redirect(this.navigateUrl);
            }, this.ironSessionConfigs);
        this.handleSignInCallback = (redirectTo = this.config.baseUrl)=>(0, $gpzKJ$ironsessionnext.withIronSessionApiRoute)(async (request, response)=>{
                const nodeClient = this.createNodeClient(request);
                if (request.url) {
                    await nodeClient.handleSignInCallback(`${this.config.baseUrl}${request.url}`);
                    await this.storage?.save();
                    response.redirect(redirectTo);
                }
            }, this.ironSessionConfigs);
        this.handleSignOut = (redirectUri = this.config.baseUrl)=>(0, $gpzKJ$ironsessionnext.withIronSessionApiRoute)(async (request, response)=>{
                const nodeClient = this.createNodeClient(request);
                await nodeClient.signOut(redirectUri);
                request.session.destroy();
                await this.storage?.save();
                if (this.navigateUrl) response.redirect(this.navigateUrl);
            }, this.ironSessionConfigs);
        this.handleUser = (configs)=>this.withLogtoApiRoute((request, response)=>{
                response.json(request.user);
            }, configs);
        this.handleAuthRoutes = (configs)=>(request, response)=>{
                const { action: action  } = request.query;
                if (action === "sign-in") return this.handleSignIn()(request, response);
                if (action === "sign-in-callback") return this.handleSignInCallback()(request, response);
                if (action === "sign-out") return this.handleSignOut()(request, response);
                if (action === "user") return this.handleUser(configs)(request, response);
                response.status(404).end();
            };
        this.withLogtoApiRoute = (handler, config = {})=>(0, $gpzKJ$ironsessionnext.withIronSessionApiRoute)(async (request, response)=>{
                const user = await this.getLogtoUserFromRequest(request, config);
                // eslint-disable-next-line @silverhand/fp/no-mutating-methods
                Object.defineProperty(request, "user", {
                    enumerable: true,
                    get: ()=>user
                });
                return handler(request, response);
            }, this.ironSessionConfigs);
        this.withLogtoSsr = (handler, configs = {})=>(0, $gpzKJ$ironsessionnext.withIronSessionSsr)(async (context)=>{
                const user = await this.getLogtoUserFromRequest(context.req, configs);
                // eslint-disable-next-line @silverhand/fp/no-mutating-methods
                Object.defineProperty(context.req, "user", {
                    enumerable: true,
                    get: ()=>user
                });
                return handler(context);
            }, this.ironSessionConfigs);
    }
    createNodeClient(request) {
        this.storage = new (0, $f80b46e338398246$export$2e2bcd8739ae039)(request);
        return new (0, ($parcel$interopDefault($gpzKJ$logtonode)))(this.config, {
            storage: this.storage,
            navigate: (url)=>{
                this.navigateUrl = url;
            }
        });
    }
    get ironSessionConfigs() {
        return {
            cookieName: `logto:${this.config.appId}`,
            password: this.config.cookieSecret,
            cookieOptions: {
                secure: this.config.cookieSecure,
                maxAge: 1209600
            }
        };
    }
    async getLogtoUserFromRequest(request, configs) {
        const nodeClient = this.createNodeClient(request);
        return nodeClient.getContext(configs);
    }
}
